### PR TITLE
Use FastCat in PT Concat implementation

### PR DESCRIPTION
**Pull request by zoranzhao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106518

Differential Revision: [D48081898](https://our.internmc.facebook.com/intern/diff/D48081898)

### CODE REVIEW

**Review by houseroad — COMMENTED** [review]
Fix the lint?

**Review by houseroad — COMMENTED** [review]
Since this change is needed by internal use case first, let's create a diff then export to github? That will be faster to land in fbcode. Otherwise, we may need to wait one day or more to get it landed internally.

**Comment by zoranzhao** [comment]
@zoranzhao has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48081898).

**Comment by zoranzhao** [comment]
> Since this change is needed by internal use case first, let's create a diff then export to github? That will be faster to land in fbcode. Otherwise, we may need to wait one day or more to get it landed internally.

@houseroad I have created a diff in fbcode first [here](https://www.internalfb.com/diff/D48082741)

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -564,10 +616,24 @@ TORCH_IMPL_FUNC(cat_out_cpu)
 
   auto materialized = tensors.materialize();
 
-  // fast path for single thread when both inputs and result are contiguous and not empty
   bool use_serial_kernel = result.numel() < at::internal::GRAIN_SIZE || at::get_num_threads() == 1;
   ScalarType dtype = materialized[valid].get().scalar_type();
   bool serial_dtype = at::isFloatingType(dtype);
+
```
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp:622]: This seems to get applied even when not single thread?

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -549,6 +549,58 @@ std::vector<Tensor> broadcast_tensors(TensorList tensors) {
...
 
+static void fastCatOutDim0(const Tensor& out, const MaterializedITensorListRef& inputs) {
+  auto outBytes = out.nbytes();
+  char* dataPtr = reinterpret_cast<char*>(out.data_ptr());
+  size_t totalBytes = 0;
+  for (const Tensor& input : inputs) {
+    if (cat_should_skip_tensor(input)) {
+      continue;
+    }
+    TORCH_CHECK(outBytes >= totalBytes);
```
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp:560]: This error check... seems kind of pointless? You are checking that the base of the range you are going to copy into is in bounds... but that's not what you really care about, you care about the end no?

**Review by ezyang — CHANGES_REQUESTED** [review]
You need perf numbers. You are applying this fast path unconditionally even when the input is large enough to admit multithreading, need to show that it is still profitable in this case.

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -564,11 +616,25 @@ TORCH_IMPL_FUNC(cat_out_cpu)
...
   bool use_serial_kernel = result.numel() < at::internal::GRAIN_SIZE || at::get_num_threads() == 1;
   ScalarType dtype = materialized[valid].get().scalar_type();
-  bool serial_dtype = at::isFloatingType(dtype);
-  if (use_serial_kernel && all_contiguous && all_same_dtype && serial_dtype) {
+  bool serial_dtype_float = at::isFloatingType(dtype);
+  bool serial_dtype_int = at::isIntegralType(dtype, /*includeBool=*/true);
+  // fast path for single thread when both inputs and result are contiguous and
+  // not empty, and concat dim is 0 or 1
+  if (use_serial_kernel && all_contiguous && all_same_dtype &&
+      (MemoryFormat::Contiguous == memory_format) && (serial_dtype_int || serial_dtype_float)) {
```
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp:626]: Actually, why do you need to know the dtype at all? lol

**Comment by ezyang** [comment]
Switching to single threaded only satisfies one major question I had. The rest is just the benchmarks.

File: aten/src/ATen/native/TensorShape.cpp (hunk 3 above)
Comment (houseroad) [reply aten/src/ATen/native/TensorShape.cpp:626]: Just to double check the complex types, their storage are also contiguous, so safe to use memcpy to copy over?

**Comment by houseroad** [comment]
Good point, benchmark data is necessary, we can have small tensor benchmark + real case data. E2E QPS improvement is also useful here. Maybe we can attach that in the internal diff test plan.

**Pull request closed without merging**